src/com/amazon/corretto/crypto/provider/AesCbcSpi.java

```diff
@@ -0,0 +1,704 @@
+                            inputLen,
+                            outputDirect,
+                            outputArray,
+                            outputOffset));
+          } else {
+            result =
+                nInitUpdateFinal(
+                    opMode,
+                    padding,
+                    key,
+                    key.length,
+                    iv,
+                    ctxContainer,
+                    0,
+                    true,
+                    inputDirect,
+                    inputArray,
+                    inputOffset,
+                    inputLen,
+                    outputDirect,
+                    outputArray,
+                    outputOffset);
+            nativeCtx = new NativeEvpCipherCtx(ctxContainer[0]);
+          }
+        } else {
+          // Cipher is in UPDATE state, which means update was called at least once, and it needs to
+          // save the context. No need to call registerMess since the first update has already done
+          // this.
+          result =
+              nativeCtx.use(
+                  ctxPtr ->
+                      nUpdateFinal(
+                          ctxPtr,
+                          true,
+                          inputDirect,
+                          inputArray,
+                          inputOffset,
+                          inputLen,
+                          unprocessedInput,
+                          outputDirect,
+                          outputArray,
+                          outputOffset));
+        }
+      } else {
+        // Don't need to save the context
+        final long ctxPtr = nativeCtx == null ? 0 : nativeCtx.take();
+        nativeCtx = null;
+        if (cipherState == CipherState.INITIALIZED) {
+          result =
+              nInitUpdateFinal(
+                  opMode,
+                  padding,
+                  key,
+                  key.length,
+                  iv,
+                  null,
+                  ctxPtr,
+                  false,
+                  inputDirect,
+                  inputArray,
+                  inputOffset,
+                  inputLen,
+                  outputDirect,
+                  outputArray,
+                  outputOffset);
+        } else {
+          // Cipher is in UPDATE state and don't need to save the context
+          result =
+              nUpdateFinal(
+                  ctxPtr,
+                  false,
+                  inputDirect,
+                  inputArray,
+                  inputOffset,
+                  inputLen,
+                  unprocessedInput,
+                  outputDirect,
+                  outputArray,
+                  outputOffset);
+        }
+      }
+
+      cipherState = CipherState.INITIALIZED;
+      unprocessedInput = 0;
+
+      return result;
+
+    } catch (final Exception e) {
+      cipherState = CipherState.NEEDS_INITIALIZATION;
+      saveNativeContextIfNeeded(ctxContainer[0]);
```

Comment:
same question as earlier about resetting ctx on state reset

Reply:
Here, we should either clean up the EVP_CIPHER_CTX by making a JNI call, or register it with the Janitor for clean up. I'm open to changing the behavior to clean up.